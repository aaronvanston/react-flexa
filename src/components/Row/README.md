# Row

A component used to display and position children columns within a grid system.

## Properties

View detailed property information in [Row props](https://github.com/aaronvanston/react-flexa#row-props) within the react-flexa readme.

- `gutter`: Sets the width of the gutter to be used between columns.
- `display`: This defines a flex container; inline or block.
- `flexDirection`: This establishes the main-axis of the flex container.
- `flexWrap`: This enables `Col` elements to wrap onto a new line.
- `justifyContent`: This defines the alignment along the main axis.
- `alignItems`: This defines how flex items are laid out along the cross axis on the current line.
- `alignContent`: This aligns a flex container's lines within when there is extra space in the cross-axis.

Flexbox descriptions sourced from [CSS-Tricks](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)


## Usage

### Basic

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row justifyContent="center" alignItems="baseline">
    ...
  </Row>
);
```

### Responsive

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row justifyContent={{ sm: "center", lg: "flex-end" }} gutter="16px">
    ...
  </Row>
);
```

Read more about react-flexa [responsive api using objects.](https://github.com/aaronvanston/react-flexa#responsive-api-using-objects)
