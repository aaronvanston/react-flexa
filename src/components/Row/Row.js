import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { themeProvider } from '../../theme';
import { filterProps, mediaQuery, gutter, CSSProperty } from '../../helpers';

const Row = styled(props =>
  React.createElement(props.elementType, filterProps(props, Row.propTypes)),
)`
  // Initial component property
  box-sizing: border-box;

  ${props => themeProvider.breakpointsKeys(props).map(breakpoint => mediaQuery(props)[breakpoint]`
    // Generate gutter
    ${gutter.row(props, breakpoint)}

    // Responsive Flexbox properties
    ${CSSProperty(props, breakpoint, 'display')}
    ${CSSProperty(props, breakpoint, 'flex-direction')}
    ${CSSProperty(props, breakpoint, 'flex-wrap')}
    ${CSSProperty(props, breakpoint, 'justify-content')}
    ${CSSProperty(props, breakpoint, 'align-items')}
    ${CSSProperty(props, breakpoint, 'align-content')}
  `)}
`;

Row.defaultProps = {
  display: 'flex',
  flexDirection: 'row',
  flexWrap: 'wrap',
  justifyContent: 'flex-start',
  alignItems: 'flex-start',
  alignContent: 'flex-start',
  elementType: 'div',
};

const displayOptions = ['flex', 'inline-flex'];
const flexDirectionOptions = ['row', 'row-reverse', 'column', 'column-reverse'];
const flexWrapOptions = ['nowrap', 'wrap', 'wrap-reverse'];
const justifyContentOptions = ['flex-start', 'flex-end', 'center', 'space-between', 'space-around'];
const alignItemsOptions = ['flex-start', 'flex-end', 'center', 'baseline', 'stretch'];
const alignContentOptions = ['flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'stretch'];

Row.propTypes = {
  gutter: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.shape({
      xs: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      sm: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      md: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      lg: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
    }),
  ]),

  display: PropTypes.oneOfType([
    PropTypes.oneOf(displayOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(displayOptions),
      sm: PropTypes.oneOf(displayOptions),
      md: PropTypes.oneOf(displayOptions),
      lg: PropTypes.oneOf(displayOptions),
    }),
  ]),

  flexDirection: PropTypes.oneOfType([
    PropTypes.oneOf(flexDirectionOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(flexDirectionOptions),
      sm: PropTypes.oneOf(flexDirectionOptions),
      md: PropTypes.oneOf(flexDirectionOptions),
      lg: PropTypes.oneOf(flexDirectionOptions),
    }),
  ]),

  flexWrap: PropTypes.oneOfType([
    PropTypes.oneOf(flexWrapOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(flexWrapOptions),
      sm: PropTypes.oneOf(flexWrapOptions),
      md: PropTypes.oneOf(flexWrapOptions),
      lg: PropTypes.oneOf(flexWrapOptions),
    }),
  ]),

  justifyContent: PropTypes.oneOfType([
    PropTypes.oneOf(justifyContentOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(justifyContentOptions),
      sm: PropTypes.oneOf(justifyContentOptions),
      md: PropTypes.oneOf(justifyContentOptions),
      lg: PropTypes.oneOf(justifyContentOptions),
    }),
  ]),

  alignItems: PropTypes.oneOfType([
    PropTypes.oneOf(alignItemsOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(alignItemsOptions),
      sm: PropTypes.oneOf(alignItemsOptions),
      md: PropTypes.oneOf(alignItemsOptions),
      lg: PropTypes.oneOf(alignItemsOptions),
    }),
  ]),

  alignContent: PropTypes.oneOfType([
    PropTypes.oneOf(alignContentOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(alignContentOptions),
      sm: PropTypes.oneOf(alignContentOptions),
      md: PropTypes.oneOf(alignContentOptions),
      lg: PropTypes.oneOf(alignContentOptions),
    }),
  ]),

  elementType: PropTypes.string,

  children: PropTypes.node,
};

Row.displayName = 'Row';

export default Row;
