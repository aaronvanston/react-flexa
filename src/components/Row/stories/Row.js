/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { select, object } from '@storybook/addon-knobs';

import { Box } from '../../../../stories/components';
import { Row, Col } from '../../';

import {
  displayOptions,
  flexDirectionOptions,
  flexWrapOptions,
  justifyContentOptions,
  alignItemsOptions,
  alignContentOptions,
} from '../Row';

export default () => (
  <Row
    style={{ padding: '10px 10px 0 10px', border: '2px solid #8bc34a', minHeight: '90vh' }}
    gutter={object('gutter (rem)', { xs: 0.5, sm: 0.5, md: 1, lg: 1 })}
    display={select('display', displayOptions, 'flex')}
    flexDirection={select('flex-direction', flexDirectionOptions, 'row')}
    flexWrap={select('flex-wrap', flexWrapOptions, 'wrap')}
    justifyContent={select('justify-content', justifyContentOptions, 'flex-start')}
    alignItems={select('align-items', alignItemsOptions, 'flex-start')}
    alignContent={select('align-content', alignContentOptions, 'flex-start')}
  >
    <Col xs={3}><Box height="5rem">col 1</Box></Col>
    <Col xs={3}><Box>col 2</Box></Col>
    <Col xs={3}><Box>col 3</Box></Col>
    <Col xs={3}><Box>col 4</Box></Col>
    <Col xs={3}><Box>col 5</Box></Col>
    <Col xs={3}><Box height="5rem">col 6</Box></Col>
  </Row>
);
