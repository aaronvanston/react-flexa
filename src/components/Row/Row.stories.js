/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import withReadme from 'storybook-readme/with-readme';

import { Row } from './stories';
import README from './README.md';

storiesOf('Row', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .add('Default', () => <Row />);
