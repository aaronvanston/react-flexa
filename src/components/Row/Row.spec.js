import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import Row from './Row';

describe('style rendering', () => {
  test('renders corrects', () => {
    const wrapper = mount(<Row />);
    expect(wrapper).toMatchStyledComponentsSnapshot();
  });

  test('generates default css', () => {
    const tree = renderer.create(<Row />).toJSON();
    expect(tree).toHaveStyleRule('box-sizing', 'border-box');
    expect(tree).toHaveStyleRule('margin-left', 'calc(-0.5rem / 2)');
    expect(tree).toHaveStyleRule('margin-right', 'calc(-0.5rem / 2)');
    expect(tree).toHaveStyleRule('display', 'flex');
    expect(tree).toHaveStyleRule('flex-direction', 'row');
    expect(tree).toHaveStyleRule('flex-wrap', 'wrap');
    expect(tree).toHaveStyleRule('justify-content', 'flex-start');
    expect(tree).toHaveStyleRule('align-items', 'flex-start');
    expect(tree).toHaveStyleRule('align-content', 'flex-start');
  });

  test('generates custom css', () => {
    const tree = renderer.create(
      <Row
        display="inline-flex"
        flexDirection="row-reverse"
        flexWrap="nowrap"
        justifyContent="center"
        alignItems="baseline"
        alignContent="stretch"
      />,
      ).toJSON();
    expect(tree).toHaveStyleRule('display', 'inline-flex');
    expect(tree).toHaveStyleRule('flex-direction', 'row-reverse');
    expect(tree).toHaveStyleRule('flex-wrap', 'nowrap');
    expect(tree).toHaveStyleRule('justify-content', 'center');
    expect(tree).toHaveStyleRule('align-items', 'baseline');
    expect(tree).toHaveStyleRule('align-content', 'stretch');
  });
});

describe('Create element', () => {
  test('Should by default create a div element', () => {
    const wrapper = shallow(<Row />);
    expect(wrapper.dive().type()).toEqual('div');
  });

  test('Should create a custom element if provided', () => {
    const wrapper = shallow(<Row elementType="span" />);
    expect(wrapper.dive().type()).toEqual('span');
    expect(wrapper.dive().type()).not.toEqual('div');
  });
});
