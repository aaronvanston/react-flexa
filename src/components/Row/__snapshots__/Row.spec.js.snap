// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style rendering renders corrects 1`] = `
<Row
  alignContent="stretch"
  alignItems="stretch"
  display="flex"
  elementType="div"
  flexDirection="row"
  flexWrap="wrap"
  justifyContent="flex-start"
>
  <Component
    alignContent="stretch"
    alignItems="stretch"
    className="sc-bdVaJa btobKP"
    display="flex"
    elementType="div"
    flexDirection="row"
    flexWrap="wrap"
    justifyContent="flex-start"
  >
    <div
      className="sc-bdVaJa btobKP"
    />
  </Component>
</Row>
`;
